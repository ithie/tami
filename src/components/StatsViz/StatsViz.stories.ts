import type { Meta, StoryObj } from '@storybook/vue3'
import StatsViz from '@/components/StatsViz/StatsViz.vue'

const meta = {
  title: 'Components/StatsViz',
  component: StatsViz,
  render: () => ({
    components: { StatsViz },
    template: '<StatsViz />'
  }),
  parameters: {},
  args: {},
  // This component will have an automatically generated docsPage entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ['autodocs']
} satisfies Meta

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {}
