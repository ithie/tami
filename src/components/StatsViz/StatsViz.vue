<template>
  <pre>{{ labelHunger }}</pre>
  <pre>{{ labelHappiness }}</pre>
  <pre>{{ labelEnergy }}</pre>
  <pre>{{ labelAge }}</pre>
</template>

<script lang="ts" setup>
import useLabels from './composables/useLabels/useLabels'

const { labelHunger, labelHappiness, labelEnergy, labelAge } = useLabels()
</script>
