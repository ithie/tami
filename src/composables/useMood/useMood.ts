import { ref } from 'vue'
import { ACTIONS } from '@/enums/ACTIONS'
import { type IMoodAge } from '@/interfaces/IMood'
import restrictRanges from '../utils/restrictRanges'

const hunger = ref(5)
const happiness = ref(5)
const energy = ref(5)
const age = ref(1)

const restrictToMaximum = () => {
  restrictRanges(hunger)
  restrictRanges(happiness)
  restrictRanges(energy)
}

const decline = () => {
  age.value++
  hunger.value += 1
  happiness.value -= 1
  energy.value -= 1
}

const ACTION_HANDLES = {
  [ACTIONS.FEED]: () => {
    hunger.value -= 2
    happiness.value += 1
    energy.value += 2
  },
  [ACTIONS.PLAY]: () => {
    happiness.value = happiness.value + 2
    energy.value -= 1
    hunger.value += 1
  },
  [ACTIONS.SLEEP]: () => {
    hunger.value += 3
    energy.value += 2
  }
}

export default (vals?: IMoodAge) => {
  if (vals) {
    hunger.value = vals.hunger || 5
    happiness.value = vals.happiness || 5
    energy.value = vals.energy || 5
    age.value = vals.age || 1
  }

  return {
    doAction: (action: ACTIONS) => {
      decline()

      if (action !== ACTIONS.NOTHING) {
        ACTION_HANDLES[action]()
      }

      restrictToMaximum()
    },
    age,
    hunger,
    happiness,
    energy
  }
}
