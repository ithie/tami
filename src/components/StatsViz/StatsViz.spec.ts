import { describe, test, expect } from 'vitest'
import { flushPromises, mount } from '@vue/test-utils'
import StatsViz from '@/components/StatsViz/StatsViz.vue'
import useMood from '@/composables/useMood/useMood'
import { type DefineComponent } from 'vue'

describe('components/StatsViz', () => {
  test('should render', async () => {
    const wrapper = mount(StatsViz as DefineComponent)

    useMood({ hunger: 3, happiness: 9, energy: 6, age: 3 })
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
  })
  test('should render max ranges', async () => {
    const wrapper = mount(StatsViz as DefineComponent)

    useMood({ hunger: 10, happiness: 10, energy: 10, age: 1 })
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()

    useMood({ hunger: 15, happiness: 15, energy: 15, age: 1 })
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
  })
  test('should render min ranges', async () => {
    const wrapper = mount(StatsViz as DefineComponent)

    useMood({ hunger: 0, happiness: 0, energy: 0, age: 1 })
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()

    useMood({ hunger: -15, happiness: -15, energy: -15, age: 1 })
    await flushPromises()
    expect(wrapper.element).toMatchSnapshot()
  })
})
