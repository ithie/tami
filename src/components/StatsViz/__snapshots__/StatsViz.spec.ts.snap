// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/StatsViz > should render 1`] = `
<div
  data-v-app=""
>
  
  <pre>
    Hunger:       ███       
  </pre>
  <pre>
    Happiness:    █████████ 
  </pre>
  <pre>
    Energy:       ██████    
  </pre>
  <pre>
    Age:          3
  </pre>
  
</div>
`;

exports[`components/StatsViz > should render max ranges 1`] = `
<div
  data-v-app=""
>
  
  <pre>
    Hunger:       ██████████
  </pre>
  <pre>
    Happiness:    ██████████
  </pre>
  <pre>
    Energy:       ██████████
  </pre>
  <pre>
    Age:          1
  </pre>
  
</div>
`;

exports[`components/StatsViz > should render max ranges 2`] = `
<div
  data-v-app=""
>
  
  <pre>
    Hunger:       ███████████████
  </pre>
  <pre>
    Happiness:    ███████████████
  </pre>
  <pre>
    Energy:       ███████████████
  </pre>
  <pre>
    Age:          1
  </pre>
  
</div>
`;

exports[`components/StatsViz > should render min ranges 1`] = `
<div
  data-v-app=""
>
  
  <pre>
    Hunger:       █████     
  </pre>
  <pre>
    Happiness:    █████     
  </pre>
  <pre>
    Energy:       █████     
  </pre>
  <pre>
    Age:          1
  </pre>
  
</div>
`;

exports[`components/StatsViz > should render min ranges 2`] = `
<div
  data-v-app=""
>
  
  <pre>
    Hunger:                                
  </pre>
  <pre>
    Happiness:                             
  </pre>
  <pre>
    Energy:                                
  </pre>
  <pre>
    Age:          1
  </pre>
  
</div>
`;
